module/zfs/zfs_log.c

```diff
@@ -558,68 +602,97 @@ zfs_log_write(zilog_t *zilog, dmu_tx_t *tx, int txtype,
 
+	/*
+	 * A WR_COPIED record must fit entirely in one log block.
+	 * Large writes can use WR_NEED_COPY, which the ZIL will
+	 * split into multiple records across several log blocks
+	 * if necessary.
+	 */
+	uint64_t max_wr_copied_lr_length = zil_max_copied_data(zilog);
+	if (write_state == WR_COPIED && resid > max_wr_copied_lr_length)
+		write_state = WR_NEED_COPY;
+
 	if ((fsync_cnt = (uintptr_t)tsd_get(zfs_fsyncer_key)) != 0) {
 		(void) tsd_set(zfs_fsyncer_key, (void *)(fsync_cnt - 1));
 	}
 
-	while (resid) {
-		itx_t *itx;
-		lr_write_t *lr;
-		itx_wr_state_t wr_state = write_state;
-		ssize_t len = resid;
+	boolean_t sync = B_TRUE;
+	if (!(ioflag & (O_SYNC | O_DSYNC)) && (zp->z_sync_cnt == 0) &&
+	    (fsync_cnt == 0))
+		sync = B_FALSE;
 
-		/*
-		 * A WR_COPIED record must fit entirely in one log block.
-		 * Large writes can use WR_NEED_COPY, which the ZIL will
-		 * split into multiple records across several log blocks
-		 * if necessary.
-		 */
-		if (wr_state == WR_COPIED &&
-		    resid > zil_max_copied_data(zilog))
-			wr_state = WR_NEED_COPY;
-		else if (wr_state == WR_INDIRECT)
-			len = MIN(blocksize - P2PHASE(off, blocksize), resid);
-
-		itx = zil_itx_create(txtype, sizeof (*lr) +
-		    (wr_state == WR_COPIED ? len : 0));
-		lr = (lr_write_t *)&itx->itx_lr;
-
-		/*
-		 * For WR_COPIED records, copy the data into the lr_write_t.
-		 */
-		if (wr_state == WR_COPIED) {
-			int err;
-			DB_DNODE_ENTER(db);
-			err = dmu_read_by_dnode(DB_DNODE(db), off, len, lr + 1,
-			    DMU_READ_NO_PREFETCH);
-			if (err != 0) {
-				zil_itx_destroy(itx);
-				itx = zil_itx_create(txtype, sizeof (*lr));
-				lr = (lr_write_t *)&itx->itx_lr;
-				wr_state = WR_NEED_COPY;
-			}
-			DB_DNODE_EXIT(db);
+	itx_t *itx;
+	if (write_state == WR_NEED_COPY) {
+		itx = zfs_log_write_itx_create(
+		    0,
+		    max_wr_copied_lr_length,
+		    write_state,
+		    zp,
+		    off,
+		    resid,
+		    sync,
+		    callback,
+		    callback_data,
+		    NULL);
+		zil_itx_assign(zilog, itx, tx);
+	} else if (write_state == WR_INDIRECT) {
+		while (resid) {
+			ssize_t len =
+			    MIN(blocksize - P2PHASE(off, blocksize), resid);
+			itx = zfs_log_write_itx_create(
+			    0,
+			    max_wr_copied_lr_length,
+			    write_state,
+			    zp,
+			    off,
+			    len,
+			    sync,
+			    callback,
+			    callback_data,
+			    NULL);
+			zil_itx_assign(zilog, itx, tx);
+			off += len;
+			resid -= len;
 		}
+	} else {
+		ASSERT3S(write_state, ==, WR_COPIED);
+
+		ASSERT3U(resid, <=, zil_max_wr_copied_lr_length(zilog));
+
+		void *copied_start;
+		itx = zfs_log_write_itx_create(
+		    resid,
+		    max_wr_copied_lr_length,
+		    write_state,
+		    zp,
+		    off,
+		    resid,
+		    sync,
+		    callback,
+		    callback_data,
+		    &copied_start);
+
+		int err;
+		DB_DNODE_ENTER(db);
+		err = dmu_read_by_dnode(DB_DNODE(db), off, resid, copied_start,
+		    DMU_READ_NO_PREFETCH);
+		if (err != 0) {
+			zil_itx_free_do_not_run_callback(itx);
```

Comment:
Is there a reason we can't continue to use `zil_itx_destroy()` here?  Are the `IMPLY/ASSERT`s wrong in some way?  It seems like `zil_itx_free_do_not_run_callback()` isn't really necessary unless it's going to be used by subsequent improvements.

Reply:
`zfs_log_write_itx_create` sets the `itx_callback` member and thus `zil_itx_destroy` would be calling the callback. That's why I added `zil_itx_free_do_NOT_RUN_CALLBACK()`.